mlfinlab/portfolio_optimization/hcaa.py

```diff
@@ -102,30 +107,39 @@ def allocate(self,
         # Calculate covariance of returns or use the user specified covariance matrix
         if covariance_matrix is None:
             covariance_matrix = asset_returns.cov()
-        cov = pd.DataFrame(covariance_matrix, index=asset_names, columns=asset_names)
+        covariance_matrix = pd.DataFrame(covariance_matrix, index=asset_names, columns=asset_names)
 
-        # Calculate correlation from covariance matrix
-        corr = self._cov2corr(covariance=cov)
+        # Calculate correlations and distances from covariance matrix
+        correlation_matrix = self._cov2corr(covariance=covariance_matrix)
+        if distance_matrix is None:
+            distance_matrix = np.sqrt((1 - correlation_matrix).round(5) / 2)
+        distance_matrix = pd.DataFrame(distance_matrix, index=asset_names, columns=asset_names)
 
         # Calculate the optimal number of clusters using the Gap statistic
         if not optimal_num_clusters:
-            optimal_num_clusters = self._get_optimal_number_of_clusters(correlation=corr,
+            optimal_num_clusters = self._get_optimal_number_of_clusters(correlation=correlation_matrix,
```

Comment:
Good question. I see that here the optimal number of clusters has been computed using a distance with factor 2 instead of 1/2. It does not change anything for the metric propriety but I left that part unchanged in order not to mess up the code. Should I use the precomputed distance ?

Reply:
Yep, so we decide what is the optimal number of clusters based on distance matrix

Reply:
@WanABRO Can you change the distance matrix back to the one used by the paper? With the factor of 2. We can always allow the user to pass a custom distance matrix but I think the one with the factor of 2 performs better in financial clustering applications

Reply:
@aditya1702 Sure, that's what I have done on the PR below

Reply:
Ok